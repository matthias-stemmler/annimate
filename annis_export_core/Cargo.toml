[package]
name = "annis_export_core"
version = "0.1.0"
edition.workspace = true

[features]
serde = ["dep:serde"]

[dependencies]
csv = "1.0"
graphannis = "3.1"
graphannis-core = "3.1"
itertools = { workspace = true }
regex = { workspace = true }
serde = { workspace = true, optional = true }
thiserror = { workspace = true }
urlencoding = "2.1"

[dev-dependencies]
indoc = "2.0"
insta = { version = "1.34", features = ["serde"] }
