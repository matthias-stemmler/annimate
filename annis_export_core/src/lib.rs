use error::AnnisExportError;
use format::export;
use graphannis::{
    corpusstorage::{CacheStrategy, CorpusInfo},
    errors::{AQLError, GraphAnnisError},
};
use query::{CorpusRef, Match, MatchesPage, MatchesPaginated, Query};
use std::{io::Write, path::Path, vec};

mod error;
mod format;
mod query;
mod util;

pub use format::ExportFormat;
pub use query::QueryConfig;
pub use query::QueryLanguage;

pub struct CorpusStorage(graphannis::CorpusStorage);

impl CorpusStorage {
    pub fn from_db_dir<P>(db_dir: P) -> Result<Self, GraphAnnisError>
    where
        P: AsRef<Path>,
    {
        Ok(Self(graphannis::CorpusStorage::with_cache_strategy(
            db_dir.as_ref(),
            CacheStrategy::PercentOfFreeMemory(25.0),
            true,
        )?))
    }

    pub fn corpus_names(&self) -> Result<CorpusNames, GraphAnnisError> {
        Ok(CorpusNames(self.0.list()?.into_iter()))
    }

    pub fn export_matches<F, W>(
        &self,
        corpus_name: &str,
        aql_query: &str,
        query_config: QueryConfig,
        format: ExportFormat,
        mut out: W,
        mut on_status: F,
    ) -> Result<(), AnnisExportError>
    where
        F: FnMut(StatusEvent),
        W: Write,
    {
        // TEMP
        let node_descriptions = self
            .0
            .node_descriptions(aql_query, query_config.query_language)?
            .into_iter()
            .take(2)
            .collect();

        let corpus_ref = CorpusRef::new(&self.0, corpus_name);
        let matches_paginated =
            MatchesPaginated::new(corpus_ref, Query::new(aql_query, query_config));
        let total_count = matches_paginated.total_count()?;
        let total_count = total_count
            .try_into()
            .map_err(|_| AnnisExportError::TooManyResults(total_count))?;
        on_status(StatusEvent::Found { count: total_count });

        let exportable_matches = ExportableMatches {
            matches_paginated,
            matches_page: None,
            total_count,
        };

        export(
            format,
            exportable_matches,
            node_descriptions,
            &mut out,
            |progress| on_status(StatusEvent::Exported { progress }),
        )?;
        out.flush()?;

        Ok(())
    }

    pub fn validate_query(
        &self,
        corpus_name: &str,
        aql_query: &str,
        query_language: QueryLanguage,
    ) -> Result<QueryValidationResult, GraphAnnisError> {
        match self
            .0
            .validate_query(&[corpus_name], aql_query, query_language)
        {
            Ok(true) => Ok(QueryValidationResult::Valid),
            Ok(false) => unreachable!("Cannot occur according to docs"),
            Err(GraphAnnisError::AQLSyntaxError(err) | GraphAnnisError::AQLSemanticError(err)) => {
                Ok(QueryValidationResult::Invalid(err))
            }
            Err(err) => Err(err),
        }
    }
}

pub struct CorpusNames(vec::IntoIter<CorpusInfo>);

impl Iterator for CorpusNames {
    type Item = String;

    fn next(&mut self) -> Option<Self::Item> {
        Some(self.0.next()?.name)
    }
}

struct ExportableMatches<'a> {
    matches_paginated: MatchesPaginated<'a>,
    matches_page: Option<MatchesPage<'a>>,
    total_count: usize,
}

impl Iterator for ExportableMatches<'_> {
    type Item = Result<Match, AnnisExportError>;

    fn next(&mut self) -> Option<Self::Item> {
        loop {
            if self.matches_page.is_none() {
                self.matches_page = match self.matches_paginated.next()? {
                    Ok(page) => Some(page),
                    Err(err) => return Some(Err(err.into())),
                };
            }

            if let Some(ref mut page) = &mut self.matches_page {
                match page.next() {
                    Some(m) => return Some(m),
                    None => self.matches_page = None,
                }
            }
        }
    }

    fn size_hint(&self) -> (usize, Option<usize>) {
        (self.total_count, Some(self.total_count))
    }
}

impl ExactSizeIterator for ExportableMatches<'_> {}

#[derive(Debug)]
pub enum StatusEvent {
    Found { count: usize },
    Exported { progress: f32 },
}

#[derive(Debug)]
pub enum QueryValidationResult {
    Valid,
    Invalid(AQLError),
}
