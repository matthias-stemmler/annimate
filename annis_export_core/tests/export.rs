use annis_export_core::{
    AnnoKey, CorpusStorage, CsvExportColumn, CsvExportConfig, ExportData, ExportDataAnno,
    ExportDataText, ExportFormat, QueryLanguage,
};
use itertools::Itertools;
use serde::Serialize;
use std::fs::File;
use std::path::Path;

macro_rules! export_test {
    ($(
        $name:ident: {
            corpus_paths: $corpus_paths:expr,
            corpus_names: $corpus_names:expr,
            aql_query: $aql_query:expr,
            query_language: $query_language:expr,
            export_columns: [$($export_columns:expr,)*],
        }
    )*) => {
        $(
            #[test]
            fn $name() {
                let test_data = TestData {
                    corpus_paths: &$corpus_paths,
                    corpus_names: &$corpus_names,
                    aql_query: $aql_query,
                    query_language: {
                        use QueryLanguage::*;
                        $query_language
                    },
                    export_columns: {
                        use TestCsvExportColumn::*;
                        use TestExportData::*;
                        vec![$($export_columns,)*]
                    },
                };

                let storage = CorpusStorage::from_db_dir(tempfile::tempdir().unwrap()).unwrap();

                for corpus_path in test_data.corpus_paths {
                    storage
                        .import_corpora_from_zip(
                            File::open(Path::new(concat!(env!("CARGO_MANIFEST_DIR"), "/tests/data")).join(corpus_path)).unwrap(),
                            |_| (),
                        )
                        .unwrap();
                }

                let mut export_bytes = Vec::new();

                storage
                    .export_matches(
                        test_data.corpus_names,
                        test_data.aql_query,
                        test_data.query_language,
                        ExportFormat::Csv(CsvExportConfig {
                            columns: test_data.export_columns.clone().into_iter().map_into().collect(),
                        }),
                        &mut export_bytes,
                        |_| (),
                    )
                    .unwrap();

                let export = String::from_utf8(export_bytes).unwrap();

                insta::with_settings!(
                    {
                         info => &test_data,
                         omit_expression => true,
                    },
                    { insta::assert_snapshot!(export) }
                );
            }
        )*
    };
}

export_test! {
    subtok_segmentation_tokens: {
        corpus_paths: ["subtok.demo_relANNIS.zip"],
        corpus_names: ["subtok.demo"],
        aql_query: "pos=\"DT\" _=_ lemma",
        query_language: AQL,
        export_columns: [
            Number,
            Data(Anno(TestExportDataAnno::Corpus {
                anno_key: ("", "language"),
            })),
            Data(Anno(TestExportDataAnno::Document {
                anno_key: ("annis", "doc"),
            })),
            Data(Anno(TestExportDataAnno::MatchNode {
                anno_key: ("grammar", "lemma"),
                index: 1,
            })),
            Data(Text(TestExportDataText {
                left_context: 4,
                right_context: 4,
                segmentation: None,
            })),
        ],
    }
    subtok_segmentation_diplomatic: {
        corpus_paths: ["subtok.demo_relANNIS.zip"],
        corpus_names: ["subtok.demo"],
        aql_query: "pos=\"DT\" _=_ lemma",
        query_language: AQL,
        export_columns: [
            Number,
            Data(Anno(TestExportDataAnno::Corpus {
                anno_key: ("", "language"),
            })),
            Data(Anno(TestExportDataAnno::Document {
                anno_key: ("annis", "doc"),
            })),
            Data(Anno(TestExportDataAnno::MatchNode {
                anno_key: ("grammar", "lemma"),
                index: 1,
            })),
            Data(Text(TestExportDataText {
                left_context: 4,
                right_context: 4,
                segmentation: Some("diplomatic"),
            })),
        ],
    }
    subtok_segmentation_norm: {
        corpus_paths: ["subtok.demo_relANNIS.zip"],
        corpus_names: ["subtok.demo"],
        aql_query: "pos=\"DT\" _=_ lemma",
        query_language: AQL,
        export_columns: [
            Number,
            Data(Anno(TestExportDataAnno::Corpus {
                anno_key: ("", "language"),
            })),
            Data(Anno(TestExportDataAnno::Document {
                anno_key: ("annis", "doc"),
            })),
            Data(Anno(TestExportDataAnno::MatchNode {
                anno_key: ("grammar", "lemma"),
                index: 1,
            })),
            Data(Text(TestExportDataText {
                left_context: 4,
                right_context: 4,
                segmentation: Some("norm"),
            })),
        ],
    }
    subtok_gap: {
        corpus_paths: ["subtok.demo_relANNIS.zip"],
        corpus_names: ["subtok.demo"],
        aql_query: "pos=\"DT\" .5,5 pos=\"DT\" & lemma & lemma & #1 _=_ #3 & #2 _=_ #4",
        query_language: AQL,
        export_columns: [
            Number,
            Data(Anno(TestExportDataAnno::Corpus {
                anno_key: ("", "language"),
            })),
            Data(Anno(TestExportDataAnno::Document {
                anno_key: ("annis", "doc"),
            })),
            Data(Anno(TestExportDataAnno::MatchNode {
                anno_key: ("grammar", "lemma"),
                index: 2,
            })),
            Data(Anno(TestExportDataAnno::MatchNode {
                anno_key: ("grammar", "lemma"),
                index: 3,
            })),
            Data(Text(TestExportDataText {
                left_context: 1,
                right_context: 1,
                segmentation: None,
            })),
        ],
    }
    subtok_varying_number_of_match_nodes: {
        corpus_paths: ["subtok.demo_relANNIS.zip"],
        corpus_names: ["subtok.demo"],
        aql_query: "(pos=\"DT\" .5,5 pos=\"DT\") | pos=\"NN\"",
        query_language: AQL,
        export_columns: [
            Number,
            Data(Anno(TestExportDataAnno::Corpus {
                anno_key: ("", "language"),
            })),
            Data(Anno(TestExportDataAnno::Document {
                anno_key: ("annis", "doc"),
            })),
            Data(Anno(TestExportDataAnno::MatchNode {
                anno_key: ("grammar", "pos"),
                index: 1,
            })),
            Data(Text(TestExportDataText {
                left_context: 1,
                right_context: 1,
                segmentation: None,
            })),
        ],
    }
    subtok_multiple_corpora: {
        corpus_paths: ["subtok.demo_relANNIS.zip", "subtok.demo2_relANNIS.zip"],
        corpus_names: ["subtok.demo", "subtok.demo2"],
        aql_query: "pos=\"DT\"",
        query_language: AQL,
        export_columns: [
            Number,
            Data(Anno(TestExportDataAnno::Corpus {
                anno_key: ("", "language"),
            })),
            Data(Anno(TestExportDataAnno::Document {
                anno_key: ("annis", "doc"),
            })),
            Data(Text(TestExportDataText {
                left_context: 1,
                right_context: 1,
                segmentation: Some("diplomatic"),
            })),
        ],
    }
    pcc2: {
        corpus_paths: ["pcc2_v7_relANNIS.zip"],
        corpus_names: ["pcc2"],
        aql_query: "Sent _i_ NP & Inf-Stat & #2 _=_ #3",
        query_language: AQL,
        export_columns: [
            Number,
            Data(Anno(TestExportDataAnno::Corpus {
                anno_key: ("", "language"),
            })),
            Data(Anno(TestExportDataAnno::Document {
                anno_key: ("annis", "doc"),
            })),
            Data(Anno(TestExportDataAnno::MatchNode {
                anno_key: ("exmaralda", "Inf-Stat"),
                index: 2,
            })),
            Data(Text(TestExportDataText {
                left_context: 10,
                right_context: 10,
                segmentation: None,
            })),
        ],
    }
}

#[derive(Serialize)]
struct TestData {
    corpus_paths: &'static [&'static str],
    corpus_names: &'static [&'static str],
    aql_query: &'static str,
    query_language: QueryLanguage,
    export_columns: Vec<TestCsvExportColumn>,
}

#[derive(Clone, Serialize)]
enum TestCsvExportColumn {
    Number,
    Data(TestExportData),
}

#[derive(Clone, Serialize)]
enum TestExportData {
    Anno(TestExportDataAnno),
    Text(TestExportDataText),
}

#[derive(Clone, Serialize)]
enum TestExportDataAnno {
    Corpus {
        anno_key: (&'static str, &'static str),
    },
    Document {
        anno_key: (&'static str, &'static str),
    },
    MatchNode {
        anno_key: (&'static str, &'static str),
        index: usize,
    },
}

#[derive(Clone, Serialize)]
struct TestExportDataText {
    pub left_context: usize,
    pub right_context: usize,
    pub segmentation: Option<&'static str>,
}

impl From<TestCsvExportColumn> for CsvExportColumn {
    fn from(column: TestCsvExportColumn) -> Self {
        match column {
            TestCsvExportColumn::Number => CsvExportColumn::Number,
            TestCsvExportColumn::Data(data) => CsvExportColumn::Data(match data {
                TestExportData::Anno(TestExportDataAnno::Corpus {
                    anno_key: (ns, name),
                }) => ExportData::Anno(ExportDataAnno::Corpus {
                    anno_key: AnnoKey {
                        ns: ns.into(),
                        name: name.into(),
                    },
                }),
                TestExportData::Anno(TestExportDataAnno::Document {
                    anno_key: (ns, name),
                }) => ExportData::Anno(ExportDataAnno::Document {
                    anno_key: AnnoKey {
                        ns: ns.into(),
                        name: name.into(),
                    },
                }),
                TestExportData::Anno(TestExportDataAnno::MatchNode {
                    anno_key: (ns, name),
                    index,
                }) => ExportData::Anno(ExportDataAnno::MatchNode {
                    anno_key: AnnoKey {
                        ns: ns.into(),
                        name: name.into(),
                    },
                    index,
                }),
                TestExportData::Text(TestExportDataText {
                    left_context,
                    right_context,
                    segmentation,
                }) => ExportData::Text(ExportDataText {
                    left_context,
                    right_context,
                    segmentation: segmentation.map(|s| s.to_string()),
                }),
            }),
        }
    }
}
