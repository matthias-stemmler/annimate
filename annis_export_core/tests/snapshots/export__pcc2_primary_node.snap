---
source: annis_export_core/tests/export.rs
info:
  corpus_paths:
    - pcc2_v7_relANNIS.zip
  corpus_names:
    - pcc2
  aql_query: "Sent _i_ NP & Inf-Stat & #2 _=_ #3"
  query_language: AQL
  export_columns:
    - Number
    - Data:
        Anno:
          Corpus:
            anno_key:
              - ""
              - language
    - Data:
        Anno:
          Document:
            anno_key:
              - annis
              - doc
    - Data:
        Anno:
          MatchNode:
            anno_key:
              - exmaralda
              - Inf-Stat
            index: 2
    - Data:
        Text:
          left_context: 10
          right_context: 10
          segmentation: ~
          primary_node_indices:
            - 1
---
Number,Corpus language,Document,#3 Inf-Stat,Left context (tokens),Match (tokens),Right context (tokens)
1,German,11299,new,Feigenblatt,Die Jugendlichen,in Zossen wollen ein Musikcafé . Das forderten sie bei
2,German,11299,new,Feigenblatt Die Jugendlichen in,Zossen,wollen ein Musikcafé . Das forderten sie bei der ersten
3,German,11299,new,Feigenblatt Die Jugendlichen in Zossen wollen,ein Musikcafé,. Das forderten sie bei der ersten Zossener Runde am
4,German,11299,giv-active,Feigenblatt Die Jugendlichen in Zossen wollen ein Musikcafé .,Das,forderten sie bei der ersten Zossener Runde am Dienstagabend .
5,German,11299,giv-active,Die Jugendlichen in Zossen wollen ein Musikcafé . Das forderten,sie,bei der ersten Zossener Runde am Dienstagabend . Dass die
6,German,11299,new,in Zossen wollen ein Musikcafé . Das forderten sie bei,der ersten Zossener Runde,am Dienstagabend . Dass die Politiker der Stadt dafür Verständnis
7,German,11299,new,sie bei der ersten Zossener Runde am Dienstagabend . Dass,die Politiker,"der Stadt dafür Verständnis haben , ist löblich . Mit"
8,German,11299,giv-active,der ersten Zossener Runde am Dienstagabend . Dass die Politiker,der Stadt,"dafür Verständnis haben , ist löblich . Mit dem Treffen"
9,German,11299,giv-active,Zossener Runde am Dienstagabend . Dass die Politiker der Stadt,dafür,"Verständnis haben , ist löblich . Mit dem Treffen im"
10,German,11299,giv-active,"der Stadt dafür Verständnis haben , ist löblich . Mit",dem Treffen,im Rathaus ist somit auch ein Dialog zwischen den Generationen
11,German,11299,acc-gen,"Verständnis haben , ist löblich . Mit dem Treffen im",Rathaus,ist somit auch ein Dialog zwischen den Generationen angestoßen .
12,German,11299,acc-inf,löblich . Mit dem Treffen im Rathaus ist somit auch,ein Dialog,zwischen den Generationen angestoßen . Dass die beiden geladenen Jugendlichen
13,German,11299,acc-inf,dem Treffen im Rathaus ist somit auch ein Dialog zwischen,den Generationen,angestoßen . Dass die beiden geladenen Jugendlichen im Laufe des
14,German,11299,giv-active,somit auch ein Dialog zwischen den Generationen angestoßen . Dass,die beiden geladenen Jugendlichen,"im Laufe des Abends immer weniger zu Wort kamen ,"
15,German,11299,giv-inactive,Generationen angestoßen . Dass die beiden geladenen Jugendlichen im Laufe,des Abends,"immer weniger zu Wort kamen , war sicher keine böse"
16,German,11299,idiom,"des Abends immer weniger zu Wort kamen , war sicher",keine böse Absicht,", ärgerlich ist es trotzdem . Und aberwitzig dazu ."
17,German,11299,idiom,", ärgerlich ist es trotzdem . Und aberwitzig dazu .",Es,"entbehrt nicht der Komik , wenn sich drei Erwachsene -"
18,German,11299,idiom,es trotzdem . Und aberwitzig dazu . Es entbehrt nicht,der Komik,", wenn sich drei Erwachsene - Karola Andrae ( Bürgerbündnis/FDP"
19,German,11299,giv-inacitve,"dazu . Es entbehrt nicht der Komik , wenn sich",drei Erwachsene,"- Karola Andrae ( Bürgerbündnis/FDP ) , Susanne Michler ("
20,German,11299,acc-gen,"entbehrt nicht der Komik , wenn sich drei Erwachsene -",Karola Andrae,"( Bürgerbündnis/FDP ) , Susanne Michler ( CDU ) und"
21,German,11299,acc-gen,"der Komik , wenn sich drei Erwachsene - Karola Andrae",( Bürgerbündnis/FDP ),", Susanne Michler ( CDU ) und Joachim Zanow ("
22,German,11299,acc-gen,"sich drei Erwachsene - Karola Andrae ( Bürgerbündnis/FDP ) ,",Susanne Michler,( CDU ) und Joachim Zanow ( SPD ) -
23,German,11299,acc-gen,"Erwachsene - Karola Andrae ( Bürgerbündnis/FDP ) , Susanne Michler",( CDU ),"und Joachim Zanow ( SPD ) - darüber streiten ,"
24,German,11299,acc-gen,"( Bürgerbündnis/FDP ) , Susanne Michler ( CDU ) und",Joachim Zanow,"( SPD ) - darüber streiten , was Jugendliche wollen"
25,German,11299,acc-gen,") , Susanne Michler ( CDU ) und Joachim Zanow",( SPD ),"- darüber streiten , was Jugendliche wollen und brauchen ,"
26,German,11299,giv-inactive,"Joachim Zanow ( SPD ) - darüber streiten , was",Jugendliche,"wollen und brauchen , ohne auf die Idee zu kommen"
27,German,11299,giv-active,"und brauchen , ohne auf die Idee zu kommen ,",sie,"selbst zu fragen . Und das , obwohl sie ihnen"
28,German,11299,giv-active,", sie selbst zu fragen . Und das , obwohl",sie,ihnen gegenüber sitzen . Die Jugendlichen wurden somit zum bloßen
29,German,11299,giv-active,"sie selbst zu fragen . Und das , obwohl sie",ihnen,gegenüber sitzen . Die Jugendlichen wurden somit zum bloßen Feigenblatt
30,German,11299,giv-active,". Und das , obwohl sie ihnen gegenüber sitzen .",Die Jugendlichen,wurden somit zum bloßen Feigenblatt degradiert . Nicht über sondern
31,German,11299,idiom,sie ihnen gegenüber sitzen . Die Jugendlichen wurden somit zum,bloßen Feigenblatt,degradiert . Nicht über sondern mit ihnen hätten die Politiker
32,German,11299,giv-active,somit zum bloßen Feigenblatt degradiert . Nicht über sondern mit,ihnen,hätten die Politiker reden sollen . Damit ist eine große
33,German,11299,giv-active,bloßen Feigenblatt degradiert . Nicht über sondern mit ihnen hätten,die Politiker,reden sollen . Damit ist eine große Chance vertan .
34,German,11299,giv-inactive,ist eine große Chance vertan . Vielleicht klappt es bei,der nächsten Runde,Anfang 2002 . Dann werden auch mehr Jugendliche eingeladen .
35,German,11299,acc-inf,bei der nächsten Runde Anfang 2002 . Dann werden auch,mehr Jugendliche,eingeladen . In der Gruppe können sie sich hoffentlich mehr
36,German,11299,acc-inf,2002 . Dann werden auch mehr Jugendliche eingeladen . In,der Gruppe,können sie sich hoffentlich mehr Gehör verschaffen . Vielleicht finden
37,German,11299,giv-active,werden auch mehr Jugendliche eingeladen . In der Gruppe können,sie,sich hoffentlich mehr Gehör verschaffen . Vielleicht finden dann auch
38,German,11299,idiom,Jugendliche eingeladen . In der Gruppe können sie sich hoffentlich,mehr Gehör,verschaffen . Vielleicht finden dann auch Vertreter von PDS und
39,German,11299,acc-inf,sich hoffentlich mehr Gehör verschaffen . Vielleicht finden dann auch,Vertreter,von PDS und Gewerbeverein ihren Weg ins Rathaus . Die
40,German,11299,acc-gen,mehr Gehör verschaffen . Vielleicht finden dann auch Vertreter von,PDS,und Gewerbeverein ihren Weg ins Rathaus . Die glänzten diesmal
41,German,11299,acc-gen,verschaffen . Vielleicht finden dann auch Vertreter von PDS und,Gewerbeverein,ihren Weg ins Rathaus . Die glänzten diesmal noch mit
42,German,11299,giv-inactive,dann auch Vertreter von PDS und Gewerbeverein ihren Weg ins,Rathaus,. Die glänzten diesmal noch mit Abwesenheit .
43,German,11299,giv-active,Vertreter von PDS und Gewerbeverein ihren Weg ins Rathaus .,Die,glänzten diesmal noch mit Abwesenheit .
44,German,11299,idiom,ihren Weg ins Rathaus . Die glänzten diesmal noch mit,Abwesenheit,.
45,German,4282,idiom,Steilpass,Wunder,gibt es immer wieder ! Erst spielen die Dallgower Gemeindevertreter
46,German,4282,new,Steilpass Wunder gibt es immer wieder ! Erst spielen,die Dallgower Gemeindevertreter,so statisch und verzagt wie die deutsche Abwehrreihe der Fußballkicker
47,German,4282,acc-inf,Erst spielen die Dallgower Gemeindevertreter so statisch und verzagt wie,die deutsche Abwehrreihe der Fußballkicker,. Und dann kommt aus der Tiefe solch ein fulminanter
48,German,4282,idiom,die deutsche Abwehrreihe der Fußballkicker . Und dann kommt aus,der Tiefe,"solch ein fulminanter Steilpass , von dem man hofft ,"
49,German,4282,new,der Fußballkicker . Und dann kommt aus der Tiefe solch,ein fulminanter Steilpass,", von dem man hofft , dass die Seeburger oder"
50,German,4282,giv-active,"kommt aus der Tiefe solch ein fulminanter Steilpass , von",dem,"man hofft , dass die Seeburger oder Groß-Glienicker Mitspieler ihn"
51,German,4282,acc-sit,"aus der Tiefe solch ein fulminanter Steilpass , von dem",man,"hofft , dass die Seeburger oder Groß-Glienicker Mitspieler ihn aufnehmen"
52,German,4282,new,"ein fulminanter Steilpass , von dem man hofft , dass",die Seeburger oder Groß-Glienicker Mitspieler,ihn aufnehmen können . Ein Befreiungsschlag ist es allerdings nicht
53,German,4282,giv-active,"dem man hofft , dass die Seeburger oder Groß-Glienicker Mitspieler",ihn,"aufnehmen können . Ein Befreiungsschlag ist es allerdings nicht ,"
54,German,4282,acc-inf,dass die Seeburger oder Groß-Glienicker Mitspieler ihn aufnehmen können .,Ein Befreiungsschlag,"ist es allerdings nicht , weil es vorerst keine Gefahr"
55,German,4282,idiom,"Ein Befreiungsschlag ist es allerdings nicht , weil es vorerst",keine Gefahr,fürs Dallgower Tor gab . Die Seeburger und einige Groß-Glienicker
56,German,4282,acc-inf,"es allerdings nicht , weil es vorerst keine Gefahr fürs",Dallgower Tor,gab . Die Seeburger und einige Groß-Glienicker haben den Ball
57,German,4282,giv-inactive,weil es vorerst keine Gefahr fürs Dallgower Tor gab .,Die Seeburger,und einige Groß-Glienicker haben den Ball erst zurückgespielt und dann
58,German,4282,giv-inactive,keine Gefahr fürs Dallgower Tor gab . Die Seeburger und,einige Groß-Glienicker,haben den Ball erst zurückgespielt und dann um so drängender
59,German,4282,acc-inf,Dallgower Tor gab . Die Seeburger und einige Groß-Glienicker haben,den Ball,erst zurückgespielt und dann um so drängender wieder gefordert .
60,German,4282,giv-active,und dann um so drängender wieder gefordert . Nun sollen,sie,"zeigen , wie sie die Chance verwerten . Eine Diskussion"
61,German,4282,giv-acitve,"drängender wieder gefordert . Nun sollen sie zeigen , wie",sie,"die Chance verwerten . Eine Diskussion , wo künftig die"
62,German,4282,acc-inf,"wieder gefordert . Nun sollen sie zeigen , wie sie",die Chance,"verwerten . Eine Diskussion , wo künftig die Trainerkabine stehen"
63,German,4282,new,"sollen sie zeigen , wie sie die Chance verwerten .",Eine Diskussion,", wo künftig die Trainerkabine stehen soll , wäre in"
64,German,4282,new,"sie die Chance verwerten . Eine Diskussion , wo künftig",die Trainerkabine,"stehen soll , wäre in der jetzigen Spielsituation verheerend ."
65,German,4282,new,", wo künftig die Trainerkabine stehen soll , wäre in",der jetzigen Spielsituation,verheerend . Und eine Parallele zu den deutschen Grotten-Kickern gibt
66,German,4282,idiom,"soll , wäre in der jetzigen Spielsituation verheerend . Und",eine Parallele,zu den deutschen Grotten-Kickern gibt es immer noch . Auch
67,German,4282,giv-inactive,in der jetzigen Spielsituation verheerend . Und eine Parallele zu,den deutschen Grotten-Kickern,gibt es immer noch . Auch wenn die Spieler aus
68,German,4282,acc-aggr,den deutschen Grotten-Kickern gibt es immer noch . Auch wenn,die Spieler,"aus den verschiedenen Vereinen zusammengewürfelt sind , sie müssen sich"
69,German,4282,giv-inactive,gibt es immer noch . Auch wenn die Spieler aus,den verschiedenen Vereinen,"zusammengewürfelt sind , sie müssen sich daran gewöhnen , dass"
70,German,4282,giv-active,"wenn die Spieler aus den verschiedenen Vereinen zusammengewürfelt sind ,",sie,"müssen sich daran gewöhnen , dass sie nun in einer"
71,German,4282,acc-aggr,"Spieler aus den verschiedenen Vereinen zusammengewürfelt sind , sie müssen",sich,"daran gewöhnen , dass sie nun in einer Mannschaft """
72,German,4282,giv-active,"zusammengewürfelt sind , sie müssen sich daran gewöhnen , dass",sie,"nun in einer Mannschaft "" Döberitzer Heide "" spielen ."
73,German,4282,acc-inf,"sie müssen sich daran gewöhnen , dass sie nun in","einer Mannschaft "" Döberitzer Heide """,spielen . Und das heißt gemeinsam und nicht gegeneinander .
74,German,4282,giv-active,"in einer Mannschaft "" Döberitzer Heide "" spielen . Und",das,heißt gemeinsam und nicht gegeneinander . Ermahnungen von der Seitenlinie
75,German,4282,new,spielen . Und das heißt gemeinsam und nicht gegeneinander .,Ermahnungen,"von der Seitenlinie , miteinander fair umzugehen und sich nicht"
76,German,4282,acc-inf,Und das heißt gemeinsam und nicht gegeneinander . Ermahnungen von,der Seitenlinie,", miteinander fair umzugehen und sich nicht beim kleinsten Schubser"
77,German,4282,acc-aggr,". Ermahnungen von der Seitenlinie , miteinander fair umzugehen und",sich,"nicht beim kleinsten Schubser gegenseitig zu zerfleischen , sind normalerweise"
78,German,4282,idiom,"der Seitenlinie , miteinander fair umzugehen und sich nicht beim",kleinsten Schubser,"gegenseitig zu zerfleischen , sind normalerweise überflüssig . Vorerst allerdings"

